src/umap_.jl

```diff
@@ -1,24 +1,43 @@
 # an implementation of Uniform Manifold Approximation and Projection
 # for Dimension Reduction, L. McInnes, J. Healy, J. Melville, 2018.
 
-# NOTE: unused for now
 struct UMAP_{S <: Real, M <: AbstractMatrix{S}, N <: AbstractMatrix{S}}
     graph::M
     embedding::N
+    data::AbstractMatrix
+    knns::AbstractMatrix{<:Integer}
+    dists::AbstractMatrix{<:Real}
 
+    function UMAP_{S, M, N}(graph, embedding, data, knns, dists) where {S<:Real,
+                                                                           M<:AbstractMatrix{S},
+                                                                           N<:AbstractMatrix{S}}
+        issymmetric(graph) || isapprox(graph, graph') || error("UMAP_ constructor expected graph to be a symmetric matrix")
+        size(knns) == size(dists) || error("UMAP_ constructor expected knns and dists to have equal size")
+        new(graph, embedding, data, knns, dists)
+    end
+    # For backwards compatibility:
     function UMAP_{S, M, N}(graph, embedding) where {S<:Real,
                                                      M<:AbstractMatrix{S},
                                                      N<:AbstractMatrix{S}}
-        issymmetric(graph) || isapprox(graph, graph') || error("UMAP_ constructor expected graph to be a symmetric matrix")
-        new(graph, embedding)
+        UMAP_{S, M, N}(graph, embedding, Matrix(undef, 0, 0), Matrix{Int}(undef, 0, 0), Matrix{S}(undef, 0, 0))
```

Comment:
One thing I was trying to do was leave old constructors since `UMAP_` had been exported before, so these changes wouldn't break old code. But since the struct didn't have much of a use before this may not be a big concern.

The other thing I was considering was I have had use cases where I have data and its corresponding embedding, although the embedding was not computed from the same package. I then wanted to use some `transform` sort of functionality, and recomputing the graph of the old data would be redundant. Constructing a partially defined struct would let a user wrap data and its embedding manually, and use these for transforming new data.
(As is, this wouldn't save a whole lot of computation since the current knn_search would recompute the ApproximateKNNGraph for large matrices anyways, but this would be an internal detail that could change later). But let me know your thoughts on this!

Reply:
Maybe you could give me an example of the workflow when you'd need a partial struct? I'm not sure I understand the use case entirely.

I wouldn't worry too much about "breaking" API changes - since this package is pre-1.0, any minor version bump technically permits breaking changes. There's still a ways to go before a 1.0 release is ready and the API could be made stable.

Reply:
Ok, I won't worry about the API changes :+1:

The use case was that a third party published data and its embedding, and I had some related data of interest that I wanted to embed along with the published data/embedding. Reconsidering this (and since we are changing the struct API anyways), I think a separate `transform` method would be much better than the partial struct.